### PR TITLE
#42 - Parse education plans (Backend)

### LINKED ISSUE #42: Parse education plans (Backend)

### PR DESCRIPTION
**Pull request by ImmortalIdiot**
Added parsing of the education plans from [url](https://www.miit.ru/sveden/education/programs) for education level `бакалавриат` and `специалитет` and saving them to the database.

### CODE REVIEW

**Comment by ImmortalIdiot** [comment]
Instead of parse pdf plans we decided to parse education plans urls and save the necessary info in the database.

**Comment by ImmortalIdiot** [comment]
In addition, the structure of the [page](https://www.miit.ru/sveden/education/programs) had been changed.

**Comment by ImmortalIdiot** [comment]
Result
image

File: backend/BackendMIIT/src/main/java/com/example/BackendMIIT/parser/AnnotationProgramsParser.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+                String educationLevel = ParserUtil.getStringFromElement(row, "td[itemprop='eduLevel']");
+                String profile = ParserUtil.getStringFromElement(row, "td[itemprop='eduProf']");
+
+                if (!isValidEducationLevel(educationLevel) ||
+                        !isValidEducationForm(educationForm) ||
+                        ParserUtil.isNullOrEmpty(profile)) {
+                    continue;
+                }
+
+                String annotationLink = "https://www.miit.ru" + extractAnnotationLink(row);
```
Comment (NN171) [inline backend/BackendMIIT/src/main/java/com/example/BackendMIIT/parser/AnnotationProgramsParser.java]: You can set "https://www.miit.ru" as variable BASE_URL

File: backend/BackendMIIT/src/main/java/com/example/BackendMIIT/parser/util/ParserUtil.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+import org.jsoup.nodes.Document;
+import org.jsoup.nodes.Element;
+import org.jsoup.select.Elements;
+
+import java.io.IOException;
+
+public class ParserUtil {
+
+    public static Document getDocumentByUrl(String url) throws IOException {
+        return Jsoup.connect(url).get();
```
Comment (NN171) [inline backend/BackendMIIT/src/main/java/com/example/BackendMIIT/parser/util/ParserUtil.java]: add maxBodySize() to read all of html

Comment (NN171) [reply backend/BackendMIIT/src/main/java/com/example/BackendMIIT/parser/util/ParserUtil.java]: Also you can use SneakyThrows instead "throws IOException" (There's no difference)

File: backend/BackendMIIT/src/main/java/com/example/BackendMIIT/parser/AnnotationProgramsParser.java (hunk 1 above)
Comment (ImmortalIdiot) [reply backend/BackendMIIT/src/main/java/com/example/BackendMIIT/parser/AnnotationProgramsParser.java]: Moved to the application properties and used by @Value

File: backend/BackendMIIT/src/main/java/com/example/BackendMIIT/parser/util/ParserUtil.java (hunk 2 above)
Comment (ImmortalIdiot) [reply backend/BackendMIIT/src/main/java/com/example/BackendMIIT/parser/util/ParserUtil.java]: Used the maxBodySize(0) and SneakyThrows annotation

**Comment by ImmortalIdiot** [comment]
Java moment:
image

File: backend/BackendMIIT/src/main/java/com/example/BackendMIIT/controller/EducationPlanController.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
+package com.example.BackendMIIT.controller;
+
+import com.example.BackendMIIT.service.EducationPlanService;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping(path = "api/miit/education-plans")
```
Comment (NN171) [inline backend/BackendMIIT/src/main/java/com/example/BackendMIIT/controller/EducationPlanController.java]: Reduce to miit/plans

Comment (ImmortalIdiot) [reply backend/BackendMIIT/src/main/java/com/example/BackendMIIT/controller/EducationPlanController.java]: Edited to "api/miit/plans"

**Comment by ImmortalIdiot** [comment]
Result:
image

**Review by NN171 — APPROVED** [review]
Ok

**Pull request merged**